### PR TITLE
gcp - kms keyring filter

### PR DESCRIPTION
**Pull request by evhenyshaliov**
Use case:
`
policies:
  - name: kms_in_use
    description: |
      There are regions in projects where  Key Management Service is not in use
    resource: gcp.kms-location
    filters:
      - type: value
        key: name
        value: 'projects/cloud-custodian/locations/us-central1'
      - not:
          - type: kms-keyring
            exist: true
`

### CODE REVIEW

File: tools/c7n_gcp/c7n_gcp/resources/kms.py [hunk 1]
```diff
@@ -187,3 +189,53 @@ def get(client, resource_info):
...
+        session = local_session(self.session_factory)
+        project = session.get_default_project()
+        return super_fetch_resources(self, {'name': 'projects/{}'.format(project)})
+
+
+@KmsLocation.filter_registry.register('kms-keyring')
+class KmsLocationKmsKeyringFilter(Filter):
+
+    schema = type_schema('kms-keyring', **{
+        'exist': {'anyOf': [{'type': 'boolean'}, {'type': 'string'}]}},)
```
Comment (kapilt) [inline tools/c7n_gcp/c7n_gcp/resources/kms.py:223]: can we simplify and just make this a boolean, and disallow string values

File: tools/c7n_gcp/c7n_gcp/resources/kms.py [hunk 2]
```diff
@@ -187,3 +189,53 @@ def get(client, resource_info):
...
+        asset_type = "cloudkms.googleapis.com/Location"
+
+    def _fetch_resources(self, query):
+        super_fetch_resources = QueryResourceManager._fetch_resources
+        session = local_session(self.session_factory)
+        project = session.get_default_project()
+        return super_fetch_resources(self, {'name': 'projects/{}'.format(project)})
+
+
+@KmsLocation.filter_registry.register('kms-keyring')
```
Comment (kapilt) [inline tools/c7n_gcp/c7n_gcp/resources/kms.py:219]: given its already being used on a kms-location, the kms prefix on the filter feels redundant.

File: tools/c7n_gcp/c7n_gcp/resources/kms.py (hunk 1 above)
Comment (evhenyshaliov) [reply tools/c7n_gcp/c7n_gcp/resources/kms.py]: done

File: tools/c7n_gcp/c7n_gcp/resources/kms.py (hunk 2 above)
Comment (evhenyshaliov) [reply tools/c7n_gcp/c7n_gcp/resources/kms.py]: done

File: tools/c7n_gcp/c7n_gcp/resources/kms.py [hunk 3]
```diff
@@ -187,3 +189,53 @@ def get(client, resource_info):
...
+        super_fetch_resources = QueryResourceManager._fetch_resources
+        session = local_session(self.session_factory)
+        project = session.get_default_project()
+        return super_fetch_resources(self, {'name': 'projects/{}'.format(project)})
+
+
+@KmsLocation.filter_registry.register('keyring')
+class KmsLocationKmsKeyringFilter(Filter):
+
+    schema = type_schema('kms-keyring', **{
```
Comment (kapilt) [inline tools/c7n_gcp/c7n_gcp/resources/kms.py:222]: this should be keyring as well

Comment (evhenyshaliov) [reply tools/c7n_gcp/c7n_gcp/resources/kms.py]: fixed

File: tools/c7n_gcp/c7n_gcp/resources/kms.py [hunk 4]
```diff
@@ -187,3 +189,53 @@ def get(client, resource_info):
...
+        client = session.client(
+            service_name='cloudkms', version='v1', component='projects.locations.keyRings')
+        accepted_resources = []
+        expecting_exist = self.data['exist']
+
+        for resource in resources:
+            kms_key_rings = client.execute_query('list', {'parent': resource['name']})
+            kms_key_rings_exist = (kms_key_rings.get('keyRings') is not None and
+                                   len(kms_key_rings.get('keyRings')) > 0)
+            if ((isinstance(expecting_exist, bool) and kms_key_rings_exist == expecting_exist) or
```
Comment (kapilt) [inline tools/c7n_gcp/c7n_gcp/resources/kms.py:237]: It’s always a bool now

File: tools/c7n_gcp/c7n_gcp/resources/kms.py [hunk 5]
```diff
@@ -187,3 +189,53 @@ def get(client, resource_info):
...
+    def process(self, resources, event=None):
+        session = local_session(self.manager.session_factory)
+        client = session.client(
+            service_name='cloudkms', version='v1', component='projects.locations.keyRings')
+        accepted_resources = []
+        expecting_exist = self.data['exist']
+
+        for resource in resources:
+            kms_key_rings = client.execute_query('list', {'parent': resource['name']})
+            kms_key_rings_exist = (kms_key_rings.get('keyRings') is not None and
```
Comment (kapilt) [inline tools/c7n_gcp/c7n_gcp/resources/kms.py:235]: This logic could be simplified `bool(kms_key_rings.get(‘keyRings’)`

File: tools/c7n_gcp/c7n_gcp/resources/kms.py [hunk 6]
```diff
@@ -187,3 +189,53 @@ def get(client, resource_info):
...
+                                   len(kms_key_rings.get('keyRings')) > 0)
+            if ((isinstance(expecting_exist, bool) and kms_key_rings_exist == expecting_exist) or
+                    (isinstance(expecting_exist, str)
+                     and kms_key_rings_exist == distutils.util.strtobool(expecting_exist))):
```
Comment (kapilt) [inline tools/c7n_gcp/c7n_gcp/resources/kms.py:239]: Distutils is going away

File: tools/c7n_gcp/c7n_gcp/resources/kms.py (hunk 4 above)
Comment (evhenyshaliov) [reply tools/c7n_gcp/c7n_gcp/resources/kms.py]: fixed, thanks

File: tools/c7n_gcp/c7n_gcp/resources/kms.py (hunk 5 above)
Comment (evhenyshaliov) [reply tools/c7n_gcp/c7n_gcp/resources/kms.py]: fixed, thanks

**Review by kapilt — APPROVED** [review]
lgtm, thanks

**Pull request merged**